### PR TITLE
Feature/convert2grant

### LINKED ISSUE #7098: Tribes V2 - Next Iteration  09/11
### Why

**For sponsors**, Tribes is a one stop shop where sponsors can manage their open source brand, stay connected with hackers, see all of the great hackathon projects that developers have created, understand the open grants associated with their company and more! All on Gitcoin!

Incorporate and connect more of the core features to make it useful.

**For hackers/developers**, Tribes is a one stop shop on Gitcoin to learn about an organization, understand what the organization is about, communicate with sponsors, see all of the great people contributing to this org, see all of the great projects being built, find meaningful open source work (bounties or hack) that they can contribute towards, and more!

### What

In  the next iteration of Tribes, we would like to improve the following to ensure sponsors have a good experience to do the above.

- [x] **Tab Organization**  - Combine tabs (Projects & Grants)
- [ ] **Projects**  - Add truncation to each project showing  3 lines of copy, add ability to show more/show less
- [x] **Projects** - Display projects that have been converted to grants.
Left rail filter: DisplayProjects with Grants and Hackathon  Projects.
- [x] **Projects** - Build Convert to Grant Flow
- [x] **Projects** -  Denote Grants with a grant badge in the upper right hand corner for hack projects that have been converted.

Nice to have  if time permits
Break Townsquare related into a new PR
- [ ] **Tribe Townsquare** - Rename tab to Tribe Wall
- [ ]  **Tribe Townsquare** - Improve feed sorting, HMW surface top 10 posts on an org? Ideas: number of likes in the last  7 days

Not on here is improvements to people  tab. Sending a message would be ideal  but we should reconcile  with what we  are doing on user directory.

### Additional Materials

Screen Shot 2020-07-14 at 2 58 48 AM

Convert to grant flow WIP
Slice 1

### PR DESCRIPTION
**Pull request by sanchaymittal**
##### Description

**Convert to Grant feature**

Video: https://www.loom.com/share/abe0eb4176e34b599442b2fcdd733b11

1. New Winner Tag before it was a ribbon
2. New Grant Tag
3. Filters at sidebar allow the listing of project converted to grants.
4. The position change of edit button

Screenshot 2020-09-13 at 2 06 27 PM
Screenshot 2020-09-13 at 2 06 53 PM

5. Owner of the project can convert the project to grant
6. project-id is passed as an optional param with URL, preloaded the fields in grant page.

Screenshot 2020-09-13 at 2 13 40 PM
Screenshot 2020-09-13 at 2 14 01 PM

7. The project owner can visit grant page to edit it

Screenshot 2020-09-13 at 2 07 14 PM

8. The user visiting the project can jump onto the grant page to donate.

Screenshot 2020-09-13 at 2 08 06 PM

9. Tribe Project tab is also updated to show the filters and tag of winner and grant.

Screenshot 2020-09-13 at 2 21 03 PM

##### Refers/Fixes
#7098

### CODE REVIEW

File: app/grants/views.py [hunk 1]
```diff
@@ -836,8 +845,12 @@ def grant_new(request):
         'conf_time_spread': conf_time_spread(),
         'gas_advisories': gas_advisories(),
         'trusted_relayer': settings.GRANTS_OWNER_ACCOUNT,
-        'grant_types': GrantType.objects.all()
+        'grant_types': GrantType.objects.all(),
+        'data': {}
     }
+    if check_profile is not None:
+        params["data"] = project_data
```
Comment (thelostone-mc) [inline app/grants/views.py:852]: This seems un-necessary ! Add this logic to L849

**Review by thelostone-mc — CHANGES_REQUESTED** [review]
https://user-images.githubusercontent.com/19514207/87412921-2b78f800-c5fc-11ea-9509-91e40793edde.jpg

^ The screenshot  don't match the design on the ticket 
Padding / text align / button color and so on!

Could we keep the code as close to the design as possible.

File: app/grants/views.py (hunk 1 above)
Comment (test-dev-git) [reply app/grants/views.py]: done

Comment (sanchaymittal) [reply app/grants/views.py]: Done

**Review by PixelantDesign — COMMENTED** [review]
Hi @sanchaymittal lets keep the existing winner banner where it is and move grants image to the left side. What does green  Edit button do?

**Comment by test-dev-git** [comment]
> Hi @sanchaymittal lets keep the existing winner banner where it is and move grants image to the left side. What does green Edit button do?

But then it will overlap with edit button if i move the grant tag to left side. The green edit button is for project owner to edit the project.

**Comment by sanchaymittal** [comment]
> Hi @sanchaymittal lets keep the existing winner banner where it is and move grants image to the left side. What does green Edit button do?

But then it will overlap with the edit button if I move the grant tag to the left side. The green edit button is for the project owner to edit the project.

**Comment by PixelantDesign** [comment]
Just for context  -convo w @sanchaymittal -- we'll keep it like this for now until @zoek1 builds **project V2**  when we will clean the card up as part of that push. @thelostone-mc @octavioamu 

Screen Shot 2020-09-14 at 7 32 50 AM

**Comment by PixelantDesign** [comment]
@sanchaymittal could  you post a video or an grok tunnel so that we can play with the feature?

**Comment by frankchen07** [comment]
yes, would like a video or live demo - that would be perfect!

**Comment by sanchaymittal** [comment]
> yes, would like a video or live demo - that would be perfect!

Video demo: https://www.loom.com/share/abe0eb4176e34b599442b2fcdd733b11

cc: @PixelantDesign

**Comment by thelostone-mc** [comment]
- squash the migrations
- get CI going 
- @frankchen07 / @PixelantDesign  could you check + approve

Additioanlly calling out that 
https://github.com/gitcoinco/web/pull/7487 shifts the current grant creation form to a new form -> so the changes for that would have to be ported over as well ! not sure how we'd want to handle that

cc @sanchaymittal

File: app/grants/views.py [hunk 2]
```diff
@@ -754,7 +755,7 @@ def grant_new_whitelabel(request):
 
 
 @login_required
-def grant_new(request):
+def grant_new(request, project_id=None):
```
Comment (thelostone-mc) [inline app/grants/views.py:758]: Could this be a GET param instead ?

**Review by thelostone-mc — CHANGES_REQUESTED** [review]
Left a bunch of comments

File: app/dashboard/templates/dashboard/hackathon/project_page.html [hunk 3]
```diff
@@ -65,18 +65,29 @@
...
+          {% endif %} 
+
           {% if is_member %}
-            <button @click="projectModal" class="d-none d-md-block btn btn-outline-gc-blue btn-lg float-right mt-5 mr-5 px-5 py-3 font-subheader"><i class="fa fa-pencil mr-2"></i> Edit project</button>
-          {% endif %}
-          <div class="row">
-            <div class="col-12 col-md-6 col-lg-8 px-5 px-md-0">
-              {% include 'project/detail/info.html' %}
-            </div>
+            {% if project_obj.grant_url is None %}
```
[the remark this answers is not in the document]
Comment (sanchaymittal) [reply app/dashboard/templates/dashboard/hackathon/project_page.html:80]: Will do that!

File: app/dashboard/templates/dashboard/hackathon/project_page.html [hunk 4]
```diff
@@ -65,18 +65,29 @@
...
+            <b-link href="{% url 'grants:new' project_obj.id %}" class="d-none d-md-block btn btn-gc-blue btn-block font-subheader btn-lg m-2"> <img class="mr-2" src="{% static 'v2/images/grants/grants-logo-2.svg' %}" alt="Create"> CONVERT TO GRANT</b-link>
```
[the remark this answers is not in the document]
Comment (sanchaymittal) [reply app/dashboard/templates/dashboard/hackathon/project_page.html:81]: Ahh forgot ```make fix```

File: app/dashboard/templates/dashboard/hackathon/project_page.html [hunk 5]
```diff
@@ -65,18 +65,29 @@
...
-        <div id="project-details" class="col-12 pr-0 ">
+        <div id="project-details" class="pr-0 ">
+          <div class="float-right mt-5 mr-5 px-5 py-3">
+
+          {% if project_obj.grant_url %}
+            {% if is_member %}
+              <b-link href="{{ project_obj.grant_url }}" class="d-none d-md-block btn btn-gc-blue btn-block font-subheader btn-lg m-2"> <img class="mr-2" src="{% static 'v2/images/grants/grants-logo-2.svg' %}" alt="Triangles"> EDIT GRANT</b-link> 
+            {% else %} 
+              <b-link href="{{ project_obj.grant_url }}" class="d-none d-md-block btn btn-gc-blue btn-block font-subheader btn-lg m-2"> <img class="mr-2" src="{% static 'v2/images/grants/grants-logo-2.svg' %}" alt="Triangles"> DONATE TO GRANT</b-link> 
+            {% endif %}
```
[the remark this answers is not in the document]
Comment (sanchaymittal) [reply app/dashboard/templates/dashboard/hackathon/project_page.html:76]: Will do that!

**Comment by frankchen07** [comment]
how is a project denoted as a grant in our DB? is there a flag that tell us if it was converted to a grant, and when it occurs?

**Comment by sanchaymittal** [comment]
> how is a project denoted as a grant in our DB? is there a flag that tell us if it was converted to a grant, and when it occurs?

A field is created in HackathonProject DB grant_link which is a foreign Key to Grant. Once a grant is created through convert2grant flow then it updates the field with grant object.

File: app/grants/views.py (hunk 2 above)
Comment (sanchaymittal) [reply app/grants/views.py:758]: Will do that!

**Comment by frankchen07** [comment]
got it, so when the grant is created via convert2grant flow, that means the `created_on` field in the grant object is essentially when the project -> grant?, and we can find that by just joining the project model to the grants model via the foreign key?

**Comment by sanchaymittal** [comment]
> got it, so when the grant is created via convert2grant flow, that means the `created_on` field in the grant object is essentially when the project -> grant, and we can find that by just joining the project model to the grants model via the foreign key?
> 
> if yes, then LGTM, we can track the outcome of this feature

Yes, when we create the grant we save the project object to the foreign key field.

File: app/dashboard/views.py [hunk 6]
```diff
@@ -4518,18 +4532,19 @@ def get_project(request, project_id):
                 'url': member_profile.url,
                 'handle': member_profile.handle,
                 'avatar': member_profile.avatar_url
-            } for member_profile in project.profiles.all()]
+            } for member_profile in project.profiles.all()],
+            'team_members_profile': [member_profile for member_profile in project.profiles.all()]
         },
         'hackathon': hackathon_obj[0],
     }
+    return params
```
Comment (thelostone-mc) [inline app/dashboard/views.py:4540]: Why isn't this a JsonResponse ?

File: app/dashboard/models.py [hunk 7]
```diff
@@ -5092,6 +5092,12 @@ class HackathonProject(SuperModel):
     chat_channel_id = models.CharField(max_length=255, blank=True, null=True)
     winner = models.BooleanField(default=False)
     extra = JSONField(default=dict, blank=True, null=True)
+    grant_link = models.ForeignKey(
+        'grants.Grant',
+        null=True,
```
Comment (thelostone-mc) [inline app/dashboard/models.py:5097]: blank=True needed

File: app/dashboard/migrations/0151_auto_20200921_1510.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+        ('dashboard', '0150_auto_20200913_1437'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='hackathonproject',
+            name='grant_link',
+            field=models.ForeignKey(help_text='Link to grant if project is converted to grant', null=True, on_delete=django.db.models.deletion.SET_NULL, to='grants.Grant'),
+        ),
+    ]
```
Comment (thelostone-mc) [inline app/dashboard/migrations/0151_auto_20200921_1510.py:19]: ^ squash migrations

**Review by thelostone-mc — COMMENTED** [review]
left a few comments !

File: app/dashboard/views.py (hunk 6 above)
Comment (sanchaymittal) [reply app/dashboard/views.py:4540]: Using this as an internal function to share project details. The JsonResponse is returned by the function above this one **get_project()**.

File: app/dashboard/models.py (hunk 7 above)
Comment (sanchaymittal) [reply app/dashboard/models.py:5097]: Will do that!

File: app/grants/templates/grants/new.html [hunk 9]
```diff
@@ -77,24 +77,28 @@ <h5 class="mt-4">Project Information</h5>
...
+                  <input type="url" class="form__input form__input-lg " id="input-url" name="reference_url" value="{{data.project.work_url}}" placeholder="https://human.fund/" required/>
                 </div>
 
                 <div class="form__group-horizontal pb-3">
                   <label class="font-body" for="input-team_members">{% trans "Project Team members" %}</label>
                   <div class="form__select2 g-multiselect">
                     <select id="input-team_members" class="form__input form__input-lg team_members" name="team_members[]" multiple="multiple">
+                      {% for profile in data.project.team_members_profile %}
+                      <option selected="selected"  value="{{ profile.id }}">{{ profile }}</option>
+                    {% endfor %}
```
[the remark this answers is not in the document]
Comment (sanchaymittal) [reply app/grants/templates/grants/new.html:101]: done

File: app/dashboard/migrations/0151_auto_20200921_1510.py (hunk 8 above)
Comment (sanchaymittal) [reply app/dashboard/migrations/0151_auto_20200921_1510.py:19]: I squashed the migrations, this file is from the older commits won't be included. https://github.com/gitcoinco/web/blob/feature/convert2grant/app/dashboard/migrations/0149_hackathonproject_grant_link_squashed_0151_auto_20200921_1510.py

Comment (sanchaymittal) [reply app/dashboard/migrations/0151_auto_20200921_1510.py:19]: I removing the last files to fix this.

**Review by PixelantDesign — APPROVED** [review]

**Review by frankchen07 — APPROVED** [review]

File: app/dashboard/templates/dashboard/hackathon/project_page.html [hunk 10]
```diff
@@ -65,18 +65,25 @@
...
-          <div class="row">
-            <div class="col-12 col-md-6 col-lg-8 px-5 px-md-0">
-              {% include 'project/detail/info.html' %}
-            </div>
+            {% if not project_obj.grant_url %}
+            <b-link href="{% url 'grants:new' %}?project_id={{project_obj.id}}" class="d-none d-md-block btn btn-gc-blue btn-block font-subheader btn-lg m-2"> <img class="mr-2" src="{% static 'v2/images/grants/grants-logo-2.svg' %}" alt="Create"> CREATE GRANT</b-link>
+            {% endif %}
+            <button @click="projectModal" class="d-none d-md-block btn btn-outline-gc-blue btn-block font-subheader btn-lg m-2"><i class="fa fa-pencil mr-2"></i> EDIT PROJECT</button>
+          {% endif %} 
+          </div>
```
Comment (thelostone-mc) [inline app/dashboard/templates/dashboard/hackathon/project_page.html:81]: ^ fix up the indentation

**Review by thelostone-mc — APPROVED** [review]
Minor comment on indentation, otherwise LGTM
@sanchaymittal this is something I would your help as a follow up PR against zcash branch  https://gitcoincore.slack.com/archives/G01CNNGMT3K/p1602241297000100

File: app/dashboard/templates/dashboard/hackathon/project_page.html (hunk 10 above)
Comment (sanchaymittal) [reply app/dashboard/templates/dashboard/hackathon/project_page.html:80]: Done

**Comment by sanchaymittal** [comment]
> Minor comment on indentation, otherwise LGTM
> @sanchaymittal this is something I would your help as a follow up PR against zcash branch [gitcoincore.slack.com/archives/G01CNNGMT3K/p1602241297000100](https://gitcoincore.slack.com/archives/G01CNNGMT3K/p1602241297000100)

Noted!

File: app/grants/views.py [hunk 11]
```diff
@@ -1329,12 +1330,21 @@ def grant_new(request):
...
                 'url': grant.url,
             })
 
-
+    check_profile = None
+    project_data = None
+    project_id = request.GET.get('project_id', None)
+    if project_id is not None:
+        check_profile = request.user.profile if request.user.is_authenticated and hasattr(request.user, 'profile') else None
+        project_data = project(project_id)
```
Comment (octavioamu) [inline app/grants/views.py:1347]: at the beginning of the code is already fetching the user, not need to find it again, also the page is @login_required

File: app/grants/views.py [hunk 12]
```diff
@@ -1351,8 +1361,10 @@ def grant_new(request):
         'conf_time_spread': conf_time_spread(),
         'gas_advisories': gas_advisories(),
         'trusted_relayer': settings.GRANTS_OWNER_ACCOUNT,
-        'grant_types': GrantType.objects.all()
+        'grant_types': GrantType.objects.all(),
+        'data': project_data
```
Comment (octavioamu) [inline app/grants/views.py:1365]: please use a better name than data, could be the same 'project_data'

File: app/grants/views.py (hunk 11 above)
Comment (octavioamu) [reply app/grants/views.py:1347]: what need to be checked here if the users "profile.id" is on the project. team_members_profile of someone can just convert projects from others.

**Review by octavioamu — CHANGES_REQUESTED** [review]
left a few comments. the rest seems looking good

File: app/grants/views.py (hunk 12 above)
Comment (sanchaymittal) [reply app/grants/views.py:1365]: Done

File: app/grants/views.py (hunk 11 above)
Comment (sanchaymittal) [reply app/grants/views.py]: Done

**Review by octavioamu — APPROVED** [review]

**Pull request merged**